<img align="right" style="padding-left:50px;" src="figures_wk4/figures_wk4/data_cleaning.png" width=350><br>
### User Bias in Data Cleaning
For your homework assignment this week, we will explore how our treatment of our data can impact the quality of our results.

**Dataset:**
The data is a Salary Survey from AskAManager.org. It’s US-centric-ish but does allow for a range of country inputs.

A list of the corresponding survey questions can be found [here](https://www.askamanager.org/2021/04/how-much-money-do-you-make-4.html).

 

In [1]:
import pandas as pd

In [2]:
df= pd.read_csv('survey_data.csv')

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28108 entries, 0 to 28107
Data columns (total 18 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   timestamp  28108 non-null  object 
 1   q1         28108 non-null  object 
 2   q2         28033 non-null  object 
 3   q3         28107 non-null  object 
 4   q4         7273 non-null   object 
 5   q5         28108 non-null  object 
 6   q6         20793 non-null  float64
 7   q7         28108 non-null  object 
 8   q8         211 non-null    object 
 9   q9         3047 non-null   object 
 10  q10        28108 non-null  object 
 11  q11        23074 non-null  object 
 12  q12        28026 non-null  object 
 13  q13        28108 non-null  object 
 14  q14        28108 non-null  object 
 15  q15        27885 non-null  object 
 16  q16        27937 non-null  object 
 17  q17        27931 non-null  object 
dtypes: float64(1), object(17)
memory usage: 3.9+ MB


In [4]:
df.head()

timestamp     q1                             q2  \
0  4/27/2021 11:02:10  25-34   Education (Higher Education)   
1  4/27/2021 11:02:22  25-34              Computing or Tech   
2  4/27/2021 11:02:38  25-34  Accounting, Banking & Finance   
3  4/27/2021 11:02:41  25-34                     Nonprofits   
4  4/27/2021 11:02:42  25-34  Accounting, Banking & Finance   

                                         q3   q4      q5      q6   q7   q8  \
0        Research and Instruction Librarian  NaN  55,000     0.0  USD  NaN   
1  Change & Internal Communications Manager  NaN  54,600  4000.0  GBP  NaN   
2                      Marketing Specialist  NaN  34,000     NaN  USD  NaN   
3                           Program Manager  NaN  62,000  3000.0  USD  NaN   
4                        Accounting Manager  NaN  60,000  7000.0  USD  NaN   

    q9             q10             q11          q12           q13  \
0  NaN   United States   Massachusetts       Boston     5-7 years   
1  NaN  United Kingdom             NaN    Cambridge  8 - 10 years   
2  NaN              US       Tennessee  Chattanooga   2 - 4 years   
3  NaN             USA       Wisconsin    Milwaukee  8 - 10 years   
4  NaN              US  South Carolina   Greenville  8 - 10 years   

           q14              q15         q16    q17  
0    5-7 years  Master's degree       Woman  White  
1    5-7 years   College degree  Non-binary  White  
2  2 - 4 years   College degree       Woman  White  
3    5-7 years   College degree       Woman  White  
4    5-7 years   College degree       Woman  White

### Assignment
Your goal for this assignment is to observe how your data treatment during the cleaning process can skew or bias the dataset.

Before diving right in, stop and read through the questions associated with the dataset. As you can see, they are either free-form text entries or categorical selections. Knowing this, perform some exploratory data analysis (EDA) to investigate the "state" of the dataset.

[Add as many code cell below here as needs]


In [5]:
df['q1'].value_counts()

q1
25-34         12668
35-44          9908
45-54          3193
18-24          1236
55-64           994
65 or over       95
under 18         14
Name: count, dtype: int64

In [10]:
df['q7'].value_counts()

q7
USD        23410
CAD         1675
GBP         1592
EUR          646
AUD/NZD      504
Other        164
CHF           37
SEK           37
JPY           23
ZAR           16
HKD            4
Name: count, dtype: int64

In [11]:
df['q11'].value_counts()

q11
California                                 2611
New York                                   2174
Massachusetts                              1522
Texas                                      1269
Illinois                                   1213
                                           ... 
Alaska, Idaho, Oregon, Utah, Washington       1
Arizona, California, Nevada, Texas            1
Illinois, Kentucky                            1
Illinois, Wisconsin                           1
Florida, Georgia                              1
Name: count, Length: 137, dtype: int64

In [13]:
df['q13'].value_counts()

q13
11 - 20 years       9630
8 - 10 years        5381
5-7 years           4886
21 - 30 years       3645
2 - 4 years         3038
31 - 40 years        870
1 year or less       533
41 years or more     125
Name: count, dtype: int64

In [14]:
df['q14'].value_counts()

q14
11 - 20 years       6542
5-7 years           6524
2 - 4 years         6263
8 - 10 years        4984
21 - 30 years       1870
1 year or less      1500
31 - 40 years        384
41 years or more      41
Name: count, dtype: int64

In [15]:
df['q15'].value_counts()

q15
College degree                        13536
Master's degree                        8879
Some college                           2075
PhD                                    1427
Professional degree (MD, JD, etc.)     1325
High School                             643
Name: count, dtype: int64

In [16]:
df['q16'].value_counts()

q16
Woman                            21389
Man                               5502
Non-binary                         747
Other or prefer not to answer      298
Prefer not to answer                 1
Name: count, dtype: int64

In [17]:
df['q17'].value_counts()

q17
White                                                                                                                                                                                                                           23235
Asian or Asian American                                                                                                                                                                                                          1410
Black or African American                                                                                                                                                                                                         694
Another option not listed here or prefer not to answer                                                                                                                                                                            625
Hispanic, Latino, or Spanish origin                                         

In [18]:
df['q2'].value_counts()

q2
Computing or Tech                          4711
Education (Higher Education)               2466
Nonprofits                                 2420
Health care                                1899
Government and Public Administration       1893
                                           ... 
Gaming (Gambling)                             1
Regulatory Affairs- nutraceuticals            1
Manufacturing : corporate admin support       1
Real Estate Investment Support                1
Social networks                               1
Name: count, Length: 1220, dtype: int64

In [21]:
df['q4'].value_counts()

q4
Fundraising                                       20
In commercial real estate industry                10
Attorney                                           8
Public library                                     7
Librarian                                          7
                                                  ..
Internal communications for an edtech company      1
Legal counsel to a state agency                    1
Oversee all staff operations for academic unit     1
AI startup                                         1
Station Commander of an Army Recruiting Team       1
Name: count, Length: 7010, dtype: int64

In [20]:
df['q9'].value_counts()

q9
Hourly                                                                                                 4
Bonus not guaranteed                                                                                   3
Stock                                                                                                  3
10 month employee                                                                                      3
Before tax                                                                                             3
                                                                                                      ..
My income is salaried but weekly hourly expectations are 55-60 hours per week.                         1
I only work part time, so my actual take-home pay is closer to $40k/year.                              1
Additional compensation is calculated as % of my annual salary based on the company annual results     1
I'm paid minimum wage for my province               

In [22]:
df['q8'].value_counts()

q8
INR                                       11
SGD                                       11
USD                                       11
NOK                                       10
DKK                                        8
                                          ..
US Dollar                                  1
AUD & NZD are not the same currency...     1
55,000                                     1
ILS/NIS                                    1
TZS                                        1
Name: count, Length: 124, dtype: int64

**Question:** How would you describe the "state" of this dataset? Be specific and detailed in your answer. (Think paragraphs rather than sentences).

Looking at some of the value counts for our categorical selection answers, it is obvious that this was answered by experienced, white, americans that are mostly over 25 years old. There are some underrepresented populations as well, with more entries being from middle eastern/white people than from only middle eastern. This doesn't necessarily mean there is a bias here, but one consideration is the format of the questionair. All of the questions are in english, and the survey is a google form survey. These two details already narrow the field to people most likely to answer the form to be Americans and people who are fluent with technology. Looking at the second question answers, we can see that 'Computing or Tech' is the number one response with almost double the amount of entries as the second most counted response, 'Higher Education'. 

There are a few catagories that could be deleted for the sake of an ML system, like questions 4 and 9. Questions 4 and 9 have more unanswered entried than they do answered entries, and based on the value counts they are all answeres that provide context to a single specific circumstance of either pay or employment. Not enough relevent information to be worth searching through. Question 8 also has more unanswered submissions than answered, however the information provided in question 8 currency types. Question 8 has about 200 answers, while the entire dataset has over 28000 entries, making only about 0.7% of entried using question 8 to provide useful information, assuming all submissions were understood and all of the question 8 entries are accurate, not duplicates to question 7 or misunderstood answeres. 

#### The Plan

Now, it is time to plan how you will clean up the dataset. You **are not** allowed to use any machine learning technique to clean the data. (No SMOTE! No machine learning! Or anything like that!)

**Question:** Based on your EDA above, detail how you would clean up this dataset. 
Things to consider: (This is not an exhaustive list)
- Are there columns that can't be effectively cleaned? If so, why?
- Are there columns that genuinely won't have a data value?
- Does it make sense to segment the dataset based on specific columns when determining how to handle the missing values?
- Are outliers a factor in this dataset?

Remember preserving as much of the data as possible is the goal. That means dropping rows with a missing value somewhere might not be the best idea.

For simplicity sake, I would remove questions 4, 8, and 9, saving resources by not having to sort through that data that is mostly empty or just extra detail. Question 8 could be important data but seeing as less than 1% of the dataset answered question 8, the question 7 answer of 'Other" can be its own category should the need arise. 

There is nothing to do after the data has been collected to represent the under-represented. The most effective way to get data to include all parties with relevent information would be to offer the survey in different laguages and using different methods, such as a paper version. From a companies sake, a single survey connected to a csv is probably the easiest way to gather and organize a good amount of data, but with that being the only way to answer the questions, there will be bias towards white people who are tech savvy. Depending on the questions asked and information needed from the dataset, maybe those of non-white ethnicity could be flagged with a greater importance to try and balance the dataset. 

Machine learning could potentially be used to fill out any blanks in question 15, but that seems to be the only question that could coorelate with the other mandatory responses in the survey.

There are enough entries that should all be seperate, so any outliers would have to astronomical to make a notable difference on any group statistics.

The only mandatory question that doesn't have all 28108 is question 3, where someone got away with leaving their job title empty. The rest of the questions that have missing entries are not required to be filled out in the survey, so for one reason or another the users have left those questions blank. No matter the reason, I don't think it would be fair to try and predict what any of the non-mandatory answers could be as the users left them blank for a reason. If we wanted to go and fill in any blanks, the questions should have been mandatory and had a 'Prefer not to answer' and 'N/A' options, so that there would be as many entries as possible and shows that the information is important to the survey.

The only exception to this I see if for question 6, we can fill the Nan values with a 0. I think the wording of the question, or lack thereof, can lead people who don't make any sort of bonus to leave the field blank rather than putting in a 0. 

#### Implementation

Based on the plan the you described above, go ahead and clean up the dataset.

[Add as many code cell below here as needs]

In [23]:
df.head(5)

timestamp     q1                             q2  \
0  4/27/2021 11:02:10  25-34   Education (Higher Education)   
1  4/27/2021 11:02:22  25-34              Computing or Tech   
2  4/27/2021 11:02:38  25-34  Accounting, Banking & Finance   
3  4/27/2021 11:02:41  25-34                     Nonprofits   
4  4/27/2021 11:02:42  25-34  Accounting, Banking & Finance   

                                         q3   q4      q5      q6   q7   q8  \
0        Research and Instruction Librarian  NaN  55,000     0.0  USD  NaN   
1  Change & Internal Communications Manager  NaN  54,600  4000.0  GBP  NaN   
2                      Marketing Specialist  NaN  34,000     NaN  USD  NaN   
3                           Program Manager  NaN  62,000  3000.0  USD  NaN   
4                        Accounting Manager  NaN  60,000  7000.0  USD  NaN   

    q9             q10             q11          q12           q13  \
0  NaN   United States   Massachusetts       Boston     5-7 years   
1  NaN  United Kingdom             NaN    Cambridge  8 - 10 years   
2  NaN              US       Tennessee  Chattanooga   2 - 4 years   
3  NaN             USA       Wisconsin    Milwaukee  8 - 10 years   
4  NaN              US  South Carolina   Greenville  8 - 10 years   

           q14              q15         q16    q17  
0    5-7 years  Master's degree       Woman  White  
1    5-7 years   College degree  Non-binary  White  
2  2 - 4 years   College degree       Woman  White  
3    5-7 years   College degree       Woman  White  
4    5-7 years   College degree       Woman  White

In [24]:
df = df.drop(columns=['q4', 'q8', 'q9'])

In [26]:
df['q6'] = df['q6'].fillna(0)

In [27]:
df.head(5)

timestamp     q1                             q2  \
0  4/27/2021 11:02:10  25-34   Education (Higher Education)   
1  4/27/2021 11:02:22  25-34              Computing or Tech   
2  4/27/2021 11:02:38  25-34  Accounting, Banking & Finance   
3  4/27/2021 11:02:41  25-34                     Nonprofits   
4  4/27/2021 11:02:42  25-34  Accounting, Banking & Finance   

                                         q3      q5      q6   q7  \
0        Research and Instruction Librarian  55,000     0.0  USD   
1  Change & Internal Communications Manager  54,600  4000.0  GBP   
2                      Marketing Specialist  34,000     0.0  USD   
3                           Program Manager  62,000  3000.0  USD   
4                        Accounting Manager  60,000  7000.0  USD   

              q10             q11          q12           q13          q14  \
0   United States   Massachusetts       Boston     5-7 years    5-7 years   
1  United Kingdom             NaN    Cambridge  8 - 10 years    5-7 years   
2              US       Tennessee  Chattanooga   2 - 4 years  2 - 4 years   
3             USA       Wisconsin    Milwaukee  8 - 10 years    5-7 years   
4              US  South Carolina   Greenville  8 - 10 years    5-7 years   

               q15         q16    q17  
0  Master's degree       Woman  White  
1   College degree  Non-binary  White  
2   College degree       Woman  White  
3   College degree       Woman  White  
4   College degree       Woman  White

#### Reflection
Write a short reflection (400-500 words) answering the following: 
- What were the biggest issues you encountered in the messy dataset?
- How did cleaning the dataset improve its usability for machine learning?
- What would happen if we trained a model on the messy dataset vs. the cleaned one?
- Do you feel you skewed or biased the dataset while cleaning it?

I'm not sure if the purpose of this assignment was to jump through a bunch of technical hoops trying to figure out different ways to manipulate the data but based on my assesment of the data, I don't believe all of that is neccessary. Without having very specific questions to answer, it is hard to tell what data is going to be useful or not. Im not entirly sure how long it takes for a ML system to read through empty cells, but having removed 3/18 columns, removing 84324 cells, I would assume that provides some level of increased efficiency for future analysis. Considering I did not feel the need to remove any single entries, I don't believe any of the numbers would change if analysis was done on the cleaned vs the uncleaned version. After changing the NaN values in q6 to 0, any analysis done using that information would be a lot lower assuming the NaN values would have been ignored previously. 

I don't believe I have caused any notable bias. When removing question 8, I removed a lot of currencies, so those countries and people that use those currencies are all represented under "other" which does bias the overall representation of the population, but as noted above, less than 1% of entries had filled out question 8, and some of the entries were repeats of question 7. With quesion 4, any additional context doesn't change the option they picked for question 3. Question 3 has an option to fill out a profession with the 'other' option, where any main information should be provided. Any additional context could be useful if a majority or at least over half of the entires had context, then we could search for keywords and phrases to possibly identify similarities and differences between users, however about 1/4th of the entries provided additional context. With such a small portion providing additional details, we would just be creating another subset of users, catagoriziong those that did answer question 4 seperatly from the rest of the dataset, creating our own bias. Same goes for question 9, not enough people answered question 9 to justify using the information provided. About 1/8th of entries had question 9 completed, so any analysis on question 9 would only be applicable to a subset of the data, creating a bias to those entries using any finding from data using question 9.

Overall I think if I was provided the reason for the survey or what research was to be done on the dataset, more questions could be ignored or certain entries could be catagorized or flagged. Given a lack of context to the survey and data, aside from collecting the info simply to have the information, it is difficult to make any judgements on what information should be "cleaned" or altered for clarity and efficiency. All of the edits I have made have been made because they are details that could bias any findings, and provide very little information and insight if left alone.

## Deliverables
Upload your Jupyter Notebook to your GitHub repo and then provide a link to that repo in Worlclass. 